op-chain-ops/cmd/celo-migrate/main.go

```diff
@@ -421,37 +454,57 @@ func runStateMigration(newDBPath string, opts stateMigrationOptions) error {
 	}
-
+	// If the L1 starting block tag is not set, we determine it dynamically by
+	// finding the most recent final L1 block at the time of the L2 fork block.
 	if config.L1StartingBlockTag == nil {
-		l1StartBlock, err = client.BlockByNumber(context.Background(), nil)
+		// Find the L1 starting block, the L2 fork block occurs 1 minute after the last celo L1 block.
+		opts.migrationBlockTime = celoL1Head.Time + 60
+		// Wait for the L2 start time before we calculate the L1 starting block.
+		tillL2Start := int64(opts.migrationBlockTime) - time.Now().Unix()
+		if tillL2Start >= 0 {
+			time.Sleep(time.Duration(tillL2Start+1) * time.Second)
+		}
+		// MostRecentFinalizedL1BlockAtTime can a few seconds so we provide a suitably long context.
+		ctx, cancel := context.WithTimeout(context.Background(), 20*time.Second)
+		defer cancel()
+		l1StartingBlockHash, err := NewBeaconClient(opts.l1BeaconRPC, opts.l1BeaconchainURL).MostRecentFinalizedL1BlockAtTime(ctx, opts.migrationBlockTime)
 		if err != nil {
-			return fmt.Errorf("cannot fetch latest block: %w", err)
+			return fmt.Errorf("failed to fetch l1startingBlock based on L2 starting block time (%v): %w", opts.migrationBlockTime, err)
 		}
-		tag := rpc.BlockNumberOrHashWithHash(l1StartBlock.Hash(), true)
-		config.L1StartingBlockTag = (*genesis.MarshalableRPCBlockNumberOrHash)(&tag)
-	} else if config.L1StartingBlockTag.BlockHash != nil {
+		config.L1StartingBlockTag = &genesis.MarshalableRPCBlockNumberOrHash{BlockHash: &l1StartingBlockHash}
+	}
+
+	if config.L1StartingBlockTag.BlockHash != nil {
 		l1StartBlock, err = client.BlockByHash(context.Background(), *config.L1StartingBlockTag.BlockHash)
 		if err != nil {
-			return fmt.Errorf("cannot fetch block by hash: %w", err)
+			return fmt.Errorf("failed to fetch l1startingBlock by hash (%v): %w", config.L1StartingBlockTag.BlockHash, err)
 		}
 	} else if config.L1StartingBlockTag.BlockNumber != nil {
 		l1StartBlock, err = client.BlockByNumber(context.Background(), big.NewInt(config.L1StartingBlockTag.BlockNumber.Int64()))
 		if err != nil {
-			return fmt.Errorf("cannot fetch block by number: %w", err)
+			return fmt.Errorf("failed to fetch l1startingBlock by number (%v): %w", config.L1StartingBlockTag.BlockNumber, err)
 		}
 	}
 
-	// Ensure that there is a starting L1 block
-	if l1StartBlock == nil {
-		return fmt.Errorf("no starting L1 block")
+	// We want to ensure that we have sufficient time to start the sequencer
+	// before it exceeds max sequencer drift which is currently 2892s ==
+	// 48.2mins. Devops need about 7 mins to get the sequencer started after the
+	// migration script is run. We give them a 10 minute buffer in case of any
+	// issues, but if we are outside of this range we will fail here.
+	var maxSequencerDriftCelo uint64 = 2892 // Duplicate of the value defined at op-node/rollup.maxSequencerDriftCelo
+	if maxSequencerDriftCelo-(opts.migrationBlockTime-l1StartBlock.Time()) < 17*60 {
```

Comment:
Given that opts.migrationBlockTime and l1StartBlock can be decided by flags, should we check for underflows here?

Reply:
This check has been moved inside the dynamic l1 block selection code, so shouldn't be run with values provided by flags.